### DISCUSSION TITLE
GraphQL error handling in Relay

### DISCUSSION THREAD

**Discussion by gileadekelvin** [sibelius/relay-workshop-old#10] [Q&A] [answered]
I want to access error information in an application to display user friendly messages. What are the best practices for handling errors in relay? 
I should model and expose the error information as part of GraphQL schema ([source](https://relay.dev/docs/guided-tour/rendering/error-states/#accessing-errors-in-graphql-responses)) and then successfully access the payload on the client? Does this work well for both queries and mutations?
 
 Example:
 ```graphQL
type Error {
  # User friendly message
  message: String!
}

type Foo {
  bar: Result | Error
}
 ```

**Comment by sibelius** [comment, marked as answer]
I recommend modelling your business errors on schema mostly on mutation outputs

why do you need errors at queries?

**Reply by gileadekelvin** [reply]
I have a query that returns private data when the user is authorized. When the user is not authorized, I want to show this specific error message and redirect to the login page. Should this error be modeled and returned on response? Is this an expected error and therefore could it be present in the schema?

Example:
```GraphQL
interface Error {
  message: String!
}

type UnauthorizedError implements Error {
  message: String!
  code: int!
}

type User {
  id: ID!
  name: String!
}

union UserResult =
  User
  | unauthorizedError

type Query {
  me: UserResult!
}
```

And then on the client:
```GraphQL
query {
    me: {
    ... on User {
      id
      name
    }
    ... on Error {
      message
    }
    ... on UnauthorizedError {
      message
      code
    }
  }
}
```

Is this a correct approach?

**Reply by sibelius** [reply]
Union will make your GraphQL more complex

for this case I would go with a 401 status code + GraphQL errors

**Reply by gileadekelvin** [reply]
I got it. Thank you!

**Reply by istarkov** [reply]
IMO union if you would add  __typename to query (so relay compiler would generate discriminated union type) is nothing bad as instead of existance of some property like 'error' you use standart type narrowing
